### DISCUSSION TITLE
Ghostty doesn't source files like .zshrc when launched from a .command file

### DISCUSSION THREAD

**Discussion by lumap** [ghostty-org/ghostty#3030] [Feature Requests, Ideas] [answered]
When launched from a .command file on macOS (.sh equivalent), Ghostty doesn't source files like .zshrc before executing the command file. This is a bug because the default Terminal app on macOS does so, and this behavior might break some people's files

**Comment by mitchellh** [comment]
Converted to a discussion to get more details here. 

My first question is: do other terminals also do this? I assume so but good to verify.

If they do, I'm curious how they do it. I think some research is required to provide some input on this.

**Reply by lumap** [reply]
checked with iTerm2, it does load those files. I'm not sure how it does it tho :/

**Reply by notjames** [reply]
same happens with bash even with the command set to `/path/to/bash --login`
the only rc file that gets sourced is `$HOME/.bashrc`

All of my other terminals source `$HOME/.bash_profile` which in turn sources the rest of the rc files.

**Reply by simunic-cz** [reply]
I'm fairly certain Alacritty suffers from the same issue. Some time ago I tried to configure Alacritty to run `tmux attach || tmux` on startup and it just immediately exited when launched as Alacritty.app, but worked perfectly when launched from an existing terminal.<br>I guessed that it had not sourced the `~/.zshrc` when launched directly - but worked when launched from existing terminal, because it inherited the environment variables from an already initialized shell.

Now the exact same happens with Ghostty - the only difference is that Ghostty is a lot easier to debug since the window helpfully stays open when the `command` option fails to execute 😆

Unfortunately hardcoding the `--noprofile --norc` (https://github.com/ghostty-org/ghostty/blob/main/src/termio/Exec.zig#L1000) makes the `command` option somewhat troublesome.<br>For instance: the vast majority of online resources speak about adding the `export PATH="/opt/homebrew/bin:$PATH"` either to `.zshrc` or `.bash_profile` files. And since neither of those are sourced by Ghostty, then no brew installed program (such as tmux) is available in Ghostty's `command` option. There are workarounds, but none without any issues:
1. **I can use absolute path** (eg. `command = "/opt/homebrew/Cellar/tmux/3.5a/bin/tmux"`)<br>
  This locks me into a specific version - when I upgrade, I have to update the Ghostty config. This also goes directly agains the multiplatform nature of Ghostty - ability to share my Ghostty config across different machines (on my Linux machine, tmux is certainly not installed at `/opt/homebrew/Cellar/tmux/...`).<br><br>
2. **I can create a script** `startup.sh` **and run this script using the** `command` **option. In this script then manually source the** `.zshrc` **file and only then run** `tmux attach || tmux`.<br>
  But manually sourcing the profile/rc files is... well, less than ideal. No to mention that the script needs to be interpreted by something, so the interpreter needs to be hardcoded into this startup script (eg. `#!/bin/zsh` vs `#!/bin/bash`). Sourcing `.zshrc` with bash prints stderr - and the default shell on most Linux distros is not `zsh`. Hence this startup script is, again, platform specific and difficult to share between Darwin and Linux.

---

@mitchellh Since you seem to have some reason for not sourcing the profile / rc files on startup, I don't actually know of any suggestion about what to with this issue... 😅 But it is in fact at least a headache (if not an issue), since the lack of environment variables (namely extended `PATH`) greatly decreases the usefulness of the `command` option. At the very least usefulness _"in a multiplatform manner."_

**Reply by c-wicklein** [reply]
> Since you seem to have some reason for not sourcing the profile / rc files on startup, I don't actually know of any suggestion about what to with this issue... 😅 But it is in fact at least a headache (if not an issue), since the lack of environment variables (namely extended `PATH`) greatly decreases the usefulness of the `command` option. At the very least usefulness _"in a multiplatform manner."

Try instead executing a login shell and writing the command you want to run to the standard input of that shell before interaction begins.  This is how I suspect Terminal.app works, and you can effectively implement this outside of Ghostty with `expect`.. example in my earlier comments further down in this discussion.

**Reply by simunic-cz** [reply]
@c-wicklein Yes... And thank you! That's certainly getting closer. I noticed your earlier comment, but originally just skimmed over it, because something doesn't seem quite kosher about that on the first glance. But then, as you stated, it's a workaround. Still:<br>
- Neither of those scripts (the original, nor the updated alternative) are exactly "multiplatform" (at least Mac vs Linux). You're assuming the home directory is located at `/Users/...` (Mac convention). I get that the script could be probably adjusted to work on both Mac and Linux - I guess by using a few conditions detecting the current OS. But that just adds more complexity, which leads me to the second issue -<br><br>
- It just seems like a lot of ~~mental~~ shell gymnastics to achieve seemingly straightforward task: „start a shell and load the environment“. Basically, with your workaround the procedure is as follows:
   1. Start a login shell (`/usr/bin/login`), but in a manner that really makes it a non-login shell (with the `-l` flag), which then...
   2. ...starts a `/bin/bash` shell (with the flags `--noprofile --norc`), which then...
   3. ...starts a `bash` shell to run your script (`#!/usr/bin/env bash`), which locates the `zsh` shell and...
   4. ...starts a `/bin/zsh` shell to execute the command passed as a `-c` flag.

Now, call me crazy - but, at least to me, that just seems crazy 😆

**Reply by simunic-cz** [reply]
What seems a bit off to me is namely the discrepancy between what I read in the Ghostty docs and what I see in the source code. I may be (and probably am) a way off, but from what I understand:<br>
- the docs ([macOS Login Shells](https://ghostty.org/docs/help/macos-login-shells)) in essence say that... well, let me quote the very first and second to last sentence to summarize it:<br>
  > Ghostty respects the macOS traditions and will start each shell as a login shell.
  ...
  > Therefore, most shell setup should still go in the .zshrc file so that any shells, terminal or otherwise, are initialized correctly.
- while the comment in source code ([src/termio/Exec.zig#L969](https://github.com/ghostty-org/ghostty/blob/main/src/termio/Exec.zig#L969)) states that by using the `-l` flag couple lines below:
   > ...but now the shell isn't a login shell.

---

**Anyone Ghostty-knowledgeable, please, correct my observations (regarding Mac):**<br>
- When no `command` option is specified, then Ghostty starts a "non-login" `bash` shell (without profile/rc), which then \*somehow\* starts a "login" `zsh` shell (with profile/rc) and all looks and works as expected<br><br>
- When a `command` option is specified, then Ghostty starts a "non-login" `bash` shell (without profile/rc), which then directly executes the specified `command` and hence the `command` doesn't have access to the profile/rc environment

<br>

Now, the issue is, that since the `command` is executed in an earlier (effectively non-login) shell than is otherwise normally (started) available in the terminal, the `command` option is much less useful. Because if you want to use the `command` option, now you have to worry about "manually setting up the environment" before you can do anything.

If we could somehow move the execution of the `command` _"down to the very last ('login') shell"_, instead of _"replacing it"_ - it seems to me the problem would be solved.<br>**I think there's noting wrong with the chain of shells Ghostty executes on startup to finally get to the login shell, but I think the** `command` **is executed in the wrong place (too early in the chain).**

**Reply by c-wicklein** [reply]
I believe that Ghostty, like any macOS application, starts in an environment that does not benefit from the initialization of a login shell; that's just how macOS is.  So, Ghostty sensibly starts an interactive shell for the user as a login shell.  This means that Ghostty is necessarily on the hook to know *how* to start the user's login shell.  The entirety of the problem here is that when Ghostty has an alternative command to execute, that command is simply passed to `/bin/sh -c` for evaluation, and I'm suggesting that (1) passing the command to a login instance of the user's shell and (2) passing the command by writing it to that shell's standard input is consistent with what the native terminal application Terminal.app must be doing and solves the originally stated problem of this discussion.

The work-around I suggested is just that, the simplest way I know to implement this behavior for ones self right now and without modifying Ghostty, but I think this behavior should ultimately be incorporated into Ghostty.

NB that what I'm suggesting isn't *only* a change in order but a change in method, that of passing the alternative   command to the user's shell as input rather than on its command line. This reduces the amount of shell-specific knowledge Ghostty requires just to implement the feature of alternative commands to literally zero and cuts out the intermediate instance of `/bin/sh`, allowing users of a crazy, non-POSIX shell to write alternative commands for their crazy, non-POSIX shell.  As what might be a bonus, Ghostty doesn't even need to know when the alternative command  fails, because there's no failed instance of `/bin/sh` involved; once the command has been written to the shell, Ghostty just turns it over to the user for interaction and that's that.

**Reply by c-wicklein** [reply]
The Expect program I offered as a work-around is very much macOS-oriented in that it determines the user's shell by running `dscl`.

```
me@mbp16-chris ~/T/ghostty (main)> dscl . -read /Users/me UserShell
UserShell: /bin/zsh
```

If I wanted to do the same thing on Linux, I'd probably use `getent` instead of `dscl`, assuming Expect doesn't have a better way to get what is the user's shell.

**Reply by c-wicklein** [reply]
And, it probably doesn't play nice with shell integration.  Running my alternative command from a login shell is relatively important to me; shell integration is a level of functionality I'm not used to and don't necessarily notice or miss.  I'm content to think of Ghostty as a faster Terminal.app.

**Reply by c-wicklein** [reply]
@mitchellh , 

I've implemented my suggestion for how to handle the `command` and `initial-command` configuration options as entirely separate behaviors with their own parameters, `send-command` and `initial-send-command`; not changing how `command` works has made it easier to experiment without having to repeatedly recompile, so I've tried a few things:

`command` unset: ZSH runs as `-/bin/zsh`, no problems.
`command = /bin/zsh`: ZSH  runs as  `-/bin/zsh`, no problems
`command = zsh`: ZSH runs as `-/zsh`, still no problems

I'll pause here to mention that as long as Ghostty runs `zsh`, I expect things to work.  I have both `~/.zshrc` and `~/.zlogin` linked to a unified initialization script that uses a sentinel environment variable to know when it has anything to do.  Whether I'm given a login shell or not isn't a big deal.

`command = fish`: `/usr/bin/login -flp me /bin/bash --noprofile --norc -c exec -l fish` fails, game over.

This is the failure case which resulted in me noticing this discussion in the first place, and this still seems like a Ghostty problem.

```
me@Awesome ~> dscl . -read /Users/me | grep UserShell
UserShell: /bin/zsh
```

Bash is not my shell, so the environment initialization ZSH is responsible for didn't happen.  Boo.
 
Finally, `command` unset, `send-command = fish`: `-/bin/zsh` has run `fish` and I'm all set.

In this last case, I've modified Surface's initialization to queue sending the `send-command` string to whatever Surface executes, i.e. to send the command line string `fish\r` to `zsh` as regular input, which both fixes my problem and makes any command line parsing concern with `command` disappear.

**Reply by c-wicklein** [reply]
@patrik-simunic-cz I think this modification does exactly what you propose as "move the execution of the command down to the very last login shell instead of replacing it."

**Comment by noway** [comment]
I ran into the same issue. Repro steps:

1. Have `zsh` and Oh My Zsh installed

2. Create `start.command`:
```bash
#!/bin/bash
printenv ZSH
echo "^ this is ZSH env variable"
exec $SHELL
```

3. Run:
```bash
open -a /Applications/Ghostty.app ./start.command
```

4. Run:
```
open -a /Applications/iTerm.app ./start.command
```

5. Run:
```
open -a /System/Applications/Utilities/Terminal.app ./start.command
```

6. Observe differences:

<img width="762" alt="Screenshot 2024-12-27 at 5 59 08 AM" src="https://github.com/user-attachments/assets/591a607e-1e50-46db-ac45-c7f5215470dd" />
<img width="682" alt="Screenshot 2024-12-27 at 5 59 11 AM" src="https://github.com/user-attachments/assets/522d85bd-97a9-462e-9ff3-3abf56472dc6" />
<img width="682" alt="Screenshot 2024-12-27 at 5 59 14 AM" src="https://github.com/user-attachments/assets/6fa605f7-9bc2-456a-8f1c-6bfb08bb5fc6" />

**Comment by thinkyhead** [comment]
Temporary workaround for your `.command` files…
```sh
if [[ $TERM_PROGRAM == "ghostty" ]]; then
  if [[ $SHELL =~ "/zsh" ]]; then
    [[ -e ~/.zprofile     ]] && source ~/.zprofile
    [[ -e ~/.zshrc        ]] && source ~/.zshrc
  elif [[ $SHELL =~ "/bash" ]]; then
    [[ -e ~/.bash_profile ]] && source ~/.bash_profile
    [[ -e ~/.bashrc       ]] && source ~/.bashrc
  fi
fi
```

**Comment by thinkyhead** [comment]
I'm also seeing a strange related bug. After the `.command` (or `.sh`, etc.) exits Ghostty uses 100% CPU while it waits for…
```
Process exited. Press any key to close the terminal.
```

Maybe related to #3998 ?

**Comment by c-wicklein** [comment]
I don't think it make sense to speak of what shell configuration files Ghostty does or does not source given that Ghostty isn't a shell.

By default, Ghostty seems to execute the user's login shell **as a login shell**, presumably by prepending a `-` to `argv[0]`.  When Ghostty executes any other command, say as configured with the `command` directive and presumably when passed a command via `open -a`, Ghostty seems to execute that command by passing it as an additional argument to `/bin/sh -c`. 

On macOS, Terminal can also be configured to execute an alternative command at start up, but crucially, besides accepting the alternative command, Terminal also has an option as to whether or not that command is "run inside [a] shell."  When this option *is not* selected, Terminal either executes the command directly or perhaps indirectly via `/bin/sh -c`; I haven't explored which.  When this option *is* selected, Terminal executes the user's login shell **as a login shell**,  uses this instance of `$SHELL` to execute the alternative command, presumably  by writing the command string to the shell's standard input (hence why the command is echoed in the terminal), and then turns the interactive shell  over to the user.  The alternative command might still be running, or it might have completed.  If the alternative command does complete at any point, the user is left to interact with the instance of his login shell.

It would be fantastic if Ghostty had an option for executing an alternative command which worked the same way as "run inside [a] shell" does with Terminal.  One nicety of the method Terminal uses is that it's not necessary to know *how* to pass a command to a shell (although `-c` does seem to be fairly universal).   Another nicety is that if the alternative command "breaks", the user still has a working shell; the terminal doesn't simply close.   Finally, and what motivated me to search discussions, is that the method Terminal uses results in the command being evaluated in a "fully initialized shell."

My login shell is `/bin/zsh`, and my personal environment is completely "configured" by ZSH's initialization scripts.  Once I have a "fully-initialized" interactive login instance of ZSH, I generally choose to run `fish`.  With Terminal, I can simply specify an alternative command of `fish`, check "run inside [a] shell", and I'm all set.  With Ghostty, because the alternative command isn't run inside a login shell, I must instead do this...

First, configure Ghostty:
```
command = /Users/me/bin/ghostty-shell-wrapper.sh fish
```

Then, provide the command wrapper:

```
#!/usr/bin/env bash

set -euo pipefail
shopt -s nullglob

SHELL=$(dscl . -read /Users/"$(whoami)" | grep -E '^UserShell:' | awk '{ print $2 }')
exec "$SHELL" -li -c "$1" "$@"
```

If I *don't* involve the intermediate login shell, Ghostty won't be able to find `fish`.  The directory `/Users/me/macpkg/fish/3.7.1/root/bin` is added to `PATH` as part of ZSH's initialization process via [Lmod](https://lmod.readthedocs.io/en/latest/index.html).

The work-around I'm using isn't quite the same as what Terminal does, but it's good enough for now.  A work-around involving `expect(1)` would allow for replicating the behavior of Terminal by actually writing the alternative command to my login shell.

**Reply by c-wicklein** [reply]
This `expect(1)` script functions as an alternative to `/Users/me/bin/ghostty-shell-wrapper.sh` above and works by starting an interactive login shell (assuming `-li` has that effect), sends the arguments of the command to the shell as input, then turns the shell over to the user.  This more or less replicates the behavior of Terminal with "run inside shell" checked.  I would have preferred to set `argv[0]` to `-$SHELL` rather than pass `-li`, but I don't see how to do that with `spawn`.

```
#!/usr/bin/env expect

log_user 0

trap {
    set rows [stty rows]
    set cols [stty columns]
    stty rows $rows columns $cols < $spawn_out(slave,name)
} WINCH

set shell [string trim [lindex [split [exec dscl . -read /Users/[exec whoami] UserShell] ":"] 1]]
spawn $shell -li
send "$argv\r"
interact
```

**Reply by c-wicklein** [reply]
Perhaps the same idiom could be used to address #2483 ?

**Comment by tko** [comment]
Could this be a case some env vars are defined in terminal-specifc rc-files which `/etc/bashrc` and/or `/etc/zshrc` source conditionally, and there isn't one for ghostty? e.g. I see
```
[ -r "/etc/bashrc_$TERM_PROGRAM" ] && . "/etc/bashrc_$TERM_PROGRAM"
```
and
```
[ -r "/etc/zshrc_$TERM_PROGRAM" ] && . "/etc/zshrc_$TERM_PROGRAM"
```
(Took me a while to figure nix was configured in `/etc/zshrc_iTerm.app` and I needed to create `/etc/zshrc_ghostty`)

**Reply by noway** [reply]
Unfortunately, I don't think that's the case for me. 

I have all the env vars when I open Ghostty normally. I only experience the issue when I drag-and-drop a `.command` file onto Ghostty or run `open -a /Applications/Ghostty.app ./start.command`.

**Comment by thinkyhead** [comment]
Ghostty also fails to handle spaces in paths and filenames when wrapping the launch command for the OPEN Apple Event. If the file you have set Ghostty as the handler for or drop on the Ghostty application is named "Launch Me.command" or "Run Me.sh" then Ghostty will throw an error that there is no file named "Launch" or "Run."

**Reply by c-wicklein** [reply]
I would start a new discussion for problems related to file names containing spaces.

**Reply by 00-kat** [reply]
That would be #2633.

**Reply by jparise** [reply]
#10583 (coming in 1.3) addresses the "spaces in filenames" problem.

**Comment by ibdul** [comment]
this issue has a _fix_ as mentioned in https://github.com/ghostty-org/ghostty/discussions/7107#discussioncomment-12860230

**Comment by mitchellh** [comment, marked as answer]
This should be resolved with Ghostty 1.2.

**Reply by thinkyhead** [reply]
> This should be resolved with Ghostty 1.2.

Not quite there yet in Ghostty 1.2.3. It still chokes on a `.command` file with a space anywhere in the path or name.

**Reply by WalterClementsJr** [reply]
I'm on 1.2.3 and `.zshrc` suddenly not sourced, while it works on xterm & konsole

**Reply by ktkennychow** [reply]
My 1.2.3 is not sourcing `.zshrc` as well, making it quite unusable

**Reply by AWDDude** [reply]
I'm on 1.2.3 and its not sourcing my .zshrc when i launch ghostty using my streamdeck.

**Reply by tib-rijo** [reply]
Same on `1.2.3` with the Fedora 43 RPM. The file is recognised when using Ptyxis or Terminator.

One clumsy workaround is to re-login to one's own machine with `machinectl shell user@`.

**Reply by raswonders** [reply]
same here after updating to 1.2.3, .zshrc is no longer applied to my zsh.

Using `shell-integration = none` in ghostty config is a workaround for me, but I guess I'm missing on integrations then.

**Reply by aestellic** [reply]
Same issue for me on 1.2.3. `shell-integration = none` fixes it, but manually adding the shell integration code to `.zshrc` still doesn't activate shell integration.

**Reply by jparise** [reply]
Just wanted to double-check that all of these reports are specifically for the ".zshrc when launched from a .command file" case. You can 👍 to signal that, and if you're case is different, please note it in a different discussion topic.

**Reply by tib-rijo** [reply]
Do you want us to open up a new issue for the cases on Linux with starting from .desktop files or running from a GNOME keyboard shortcut?

**Reply by jparise** [reply]
@tib-rijo I think `.desktop` files are similar enough to this case that we can keep it in the same discussion. I don't know much about GNOME keyboard shortcuts, but if they use a similar launch mechanism, they can also stay here.

**Reply by jparise** [reply]
#10583 (coming in 1.3) addresses the "spaces in filenames" (e.g. `.command` files) problem for macOS.

**Reply by jparise** [reply]
#10044 (in `tip` and coming in 1.3) also improved the shell integration code in general with regard to spaces and argument handling. It sounds like most of the reports here are for 1.2.3, so it would be helpful to know if `tip` fixes those issues or if there are still more cases to cover.
